### ISSUE TITLE
Need semantic error for array with no bounds

### ISSUE THREAD

**Issue by implicitall** [lfortran/lfortran#2237]
The following short program causes a segfault. I believe there needs to be some checking of the bounds but I couldn't figure out a strategy for this.
```
program test
integer :: a(:)
a(:) = 1
end program
```

**Comment by implicitall** [comment]
Pranavchiku:i2121_1 is related to this issue.

**Comment by certik** [comment]
The error for this program should be that `a(:)` can't be used inside a program. Only for arguments of a function/subroutine.

**Comment by implicitall** [comment]
True, gfortran says "Array ‘a’ at (1) cannot have a deferred shape". But it is allowed if a has the allocatable attribute.

**Comment by certik** [comment]
Yes, allocatable and pointer attributes are allowed with `a(:)`, then it can be a local variable. So for those cases we need to give a runtime error in Debug mode, that the variable is not allocated / associated.
